src/execution/operator/aggregate/physical_window.cpp

```diff
@@ -789,21 +812,48 @@ idx_t PhysicalWindow::GetBatchIndex(ExecutionContext &context, DataChunk &chunk,
 	return lstate.batch_index;
 }
 
+bool WindowGlobalSourceState::UpdateBlockedTasks(bool blocked, InterruptState &interrupt_state) const {
+	lock_guard<mutex> guard(lock);
+	if (blocked) {
+		blocked_tasks.push_back(interrupt_state);
+	} else {
+		// The pipeline is unblocked, so flush tasks
+		for (auto &state : blocked_tasks) {
+			state.Callback();
+		}
+		blocked_tasks.clear();
+	}
+	return blocked;
+}
+
 SourceResultType PhysicalWindow::GetData(ExecutionContext &context, DataChunk &chunk,
                                          OperatorSourceInput &input) const {
 	auto &gsource = input.global_state.Cast<WindowGlobalSourceState>();
 	auto &lsource = input.local_state.Cast<WindowLocalSourceState>();
-	while (chunk.size() == 0) {
-		//	Move to the next bin if we are done.
-		while (!lsource.scanner) {
-			if (!lsource.NextPartition()) {
-				return chunk.size() > 0 ? SourceResultType::HAVE_MORE_OUTPUT : SourceResultType::FINISHED;
-			}
-		}
 
-		lsource.Scan(chunk);
-		gsource.returned += chunk.size();
+#if 1
+	//	Debugging variant
+	try {
+		while (!lsource.GetData(chunk)) {
+			TaskScheduler::GetScheduler(context.client).YieldThread();
+		}
+	} catch (...) {
+		gsource.stopped = true;
+		throw;
+	}
+#else
+	try {
+		const auto blocked = !lsource.GetData(chunk);
+		if (gsource.UpdateBlockedTasks(blocked, input.interrupt_state)) {
```

Comment:
Looking at this I suspect the problem is a race condition where T1 leaves the `GetData` function successfully (unblocking all tasks), followed by T2 blocking on old information. If T1 is returning the final chunk, this can then lead to T2 remaining blocked indefinitely as T1 will never come back to unblock T2.

If we look at e.g. the [hash join](https://github.com/duckdb/duckdb/blob/9eb0e58a02dfcc65695932a7611958994c25b65e/src/execution/operator/join/physical_hash_join.cpp#L1043) we can see how that is dealt with there - essentially *after* obtaining the global lock to block a task we check *again* if there are tasks available to execute or if we are finished.

By checking again after obtaining the lock, we prevent this race, since either (1) we block, and we are certain that we will be unblocked at some point as `lsource.GetData` returned false and it should return true in the future for another thread, or (2) we don't block after all.